javascript/ql/lib/semmle/javascript/frameworks/Spife.qll

```diff
@@ -0,0 +1,416 @@
+    RouteHandler rh;
+
+    RequestHeaderAccess() {
+      this =
+        rh.getARequestSource().ref().getAPropertyRead(["headers", "rawHeaders"]).getAPropertyRead()
+    }
+
+    override string getAHeaderName() { result = super.getPropertyName().toLowerCase() }
+
+    override RouteHandler getRouteHandler() { result = rh }
+
+    override string getKind() { result = "header" }
+  }
+
+  /**
+   * A Spife response source, that is, the response variable used by a
+   * route handler.
+   */
+  private class ReplySource extends Http::Servers::ResponseSource instanceof DataFlow::CallNode {
+    ReplySource() {
+      // const reply = require("@npm/spife/reply")
+      // reply(resp)
+      // reply.header(resp, 'foo', 'bar')
+      this = API::moduleImport(["@npm/spife/reply", "spife/reply"]).getACall() or
+      this = API::moduleImport(["@npm/spife/reply", "spife/reply"]).getAMember().getACall()
+    }
+
+    private DataFlow::SourceNode reachesHandlerReturn(DataFlow::TypeTracker t) {
+      result = this and
+      t.start()
+      or
+      exists(DataFlow::TypeTracker t2 | result = this.reachesHandlerReturn(t2).track(t2, t))
+    }
+
+    /**
+     * Gets the route handler that provides this response.
+     */
+    override RouteHandler getRouteHandler() {
+      exists(RouteHandler handler |
+        handler.(DataFlow::FunctionNode).getAReturn().getALocalSource() =
+          this.reachesHandlerReturn(DataFlow::TypeTracker::end()) and
+        result = handler
+      )
+    }
+  }
+
+  /**
+   * An HTTP header defined in a Spife response.
+   */
+  private class HeaderDefinition extends Http::ExplicitHeaderDefinition, DataFlow::MethodCallNode instanceof ReplySource {
+    HeaderDefinition() {
+      // reply.header(RESPONSE, 'Cache-Control', 'no-cache')
+      exists(DataFlow::MethodCallNode call |
+        this.ref() = call and
+        call.getMethodName() = "header" and
+        call.getNumArgument() = 3
+      )
+    }
+
+    override predicate definesHeaderValue(string headerName, DataFlow::Node headerValue) {
+      // reply.header(RESPONSE, 'Cache-Control', 'no-cache')
+      headerName = this.getNameNode().getStringValue() and
+      headerValue = this.getArgument(2)
+    }
+
+    override DataFlow::Node getNameNode() { result = this.getArgument(1) }
+
+    override RouteHandler getRouteHandler() { result = this.getRouteHandler() }
+  }
+
+  /**
+   * An invocation that sets any number of headers of the HTTP response.
+   */
+  private class MultipleHeaderDefinitions extends Http::ExplicitHeaderDefinition, DataFlow::CallNode {
+    ReplySource reply;
+
+    MultipleHeaderDefinitions() {
+      // reply.header(RESPONSE, {'Cache-Control': 'no-cache'})
+      // reply(RESPONSE, {'Cache-Control': 'no-cache'})
+      reply.ref().(DataFlow::CallNode).getCalleeName() = ["header", "reply"] and
+      reply.ref().(DataFlow::CallNode).getAnArgument().getALocalSource() instanceof
+        DataFlow::ObjectLiteralNode and
+      this = reply
```

Comment:
Multiple calls to `.ref()`, each call might not have the same value.  
Also, you won't flag `reply.header(..)` with this code.  

Also, it's the call that defines a header, not the reply.  

```suggestion
      exists(DataFlow::CallNode call | call = [reply.ref(), reply.ref().getAMethodCall("header")] | 
        call.getAnArgument().getALocalSource() instanceof DataFlow::ObjectLiteralNode and
        this = call
      )
```

(This will definitely cause the autoformatter to fail).  

Additionally. You could "save" the object-literal in a field, and then use that field in the member predicates in this class.  (Specifically `getAHeaderSource()`).